### ISSUE TITLE
Emulate H2's MERGE statement for PostgreSQL 9.5

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#4464]
H2's `MERGE` statement can be emulated in PostgreSQL 9.5:

```sql
-- H2
MERGE INTO table(col1, col2, ...)
[ KEY (...) ]
VALUES (val1, val2, ...)

-- PostgreSQL
INSERT INTO table(col1, col2, ...)
VALUES (val1, val2)
ON CONFLICT ( { explicit key columns, or implicit primary key } )
DO UPDATE SET col1 = val1, col2 = val2
```

----
See also:
- #4299

**Comment by lukaseder** [comment]
Reopening to support emulation of the `KEY` clause

**Comment by lukaseder** [comment]
Will implement `KEY` clause support later

**Comment by testark** [comment]
@lukaseder have there been any updated in the implementation of `KEY` class support? I'm badly in need of that.

**Comment by lukaseder** [comment]
Thank you for your enquiry, @testark. In fact, issue https://github.com/jOOQ/jOOQ/issues/4480 was wrongly labeled. The H2 `KEY` clause will be emulated in MariaDB, MySQL, and PostgreSQL 9.5

**Comment by testark** [comment]
@lukaseder thank you! Looking forward for the 3.8 release, then :-)

**Comment by testark** [comment]
@lukaseder just updated to jOOQ 3.8 and I can't seem to find an override of `.onDuplicateKeyUpdate` which takes a column or constraint name. Am I missing something?

**Comment by lukaseder** [comment]
The idea is that you use the `H2` syntax, which has a `KEY` clause:

```
MERGE INTO table(col1, col2, ...)
[ KEY (...) ]
VALUES (val1, val2, ...)
```

For example:

```java
        create().mergeInto(TAuthor(), TAuthor_ID(), TAuthor_FIRST_NAME(), TAuthor_LAST_NAME())
                .key(TAuthor_LAST_NAME())
                .values(3, "Lukas", "Hesse")
                .execute());
```

**Comment by testark** [comment]
@lukaseder this doesn't seems to be working as of 3.8.1. No matter what column I provide in `.key()`, the generated SQL always contains `... ON CONFLICT ("id") UPDATE ...`.

**Comment by Mcbigs** [comment]
I seem to have the same problem as @testark . `on conflict ("id")` is also stated even though I use the `key()` method.

**Comment by lukaseder** [comment]
Thanks for the hints. I'm sorry, I don't know what I was missing back then, I can confirm that this will generate an `ON CONFLICT()` clause on the primary key.

I think the best solution here would be to formally add support for `ON CONFLICT()` to the jOOQ API

**Comment by lukaseder** [comment]
With the implementation of proper `ON CONFLICT` support, the `KEY()` clause now works as well (https://github.com/jOOQ/jOOQ/issues/5297). This will be shipped in jOOQ 3.9
